packages/@ourworldindata/grapher/src/entitySelector/EntitySelector.tsx

```diff
@@ -60,162 +56,377 @@ export class EntitySelector extends React.Component<{
+
+    @computed get isMultiMode(): boolean {
         return !this.manager.canChangeEntity
     }
 
     @computed get fuzzy(): FuzzySearch<SearchableEntity> {
         return new FuzzySearch(this.searchableEntities, "name")
     }
 
-    @computed private get searchableEntities(): SearchableEntity[] {
-        return this.sortedAvailableEntities.map((name) => {
-            return { name } as SearchableEntity
-        })
-    }
-
-    @computed get searchResults(): SearchableEntity[] {
+    @computed get searchResults(): SearchableEntity[] | undefined {
         return this.searchInput
             ? this.fuzzy.search(this.searchInput)
-            : this.searchableEntities
+            : undefined
     }
 
-    @action.bound onSelect(entityName: string): void {
-        if (this.isMulti) {
+    @computed get partitionedSearchResults(): PartitionedEntities | undefined {
+        const { searchResults } = this
+
+        if (!searchResults) return undefined
+
+        const [selected, unselected] = partition(
+            searchResults.map((entity) => entity.name),
+            (name) => this.isEntitySelected(name)
+        )
+
+        return { selected, unselected }
+    }
+
+    @computed get partitionedAvailableEntities(): PartitionedEntities {
+        const [selected, unselected] = partition(
+            this.sortedAvailableEntities,
+            (name) => this.isEntitySelected(name)
+        )
+
+        return { selected, unselected }
+    }
+
+    @computed get partitionedVisibleEntities(): PartitionedEntities {
+        return (
+            this.partitionedSearchResults ?? this.partitionedAvailableEntities
+        )
+    }
+
+    @action.bound onSearchKeyDown(e: React.KeyboardEvent<HTMLElement>): void {
+        const { searchResults } = this
+        if (e.key === "Enter" && searchResults && searchResults.length > 0) {
+            this.onChange(searchResults[0].name)
+            this.clearSearchInput()
+        }
+    }
+
+    @action.bound onChange(entityName: string): void {
+        if (this.isMultiMode) {
             this.selectionArray.toggleSelection(entityName)
         } else {
             this.selectionArray.setSelectedEntities([entityName])
             if (this.props.onDismiss) this.props.onDismiss()
         }
-    }
 
-    componentDidMount(): void {
-        if (this.props.autoFocus && !isTouchDevice()) this.searchField.focus()
-    }
+        this.set({ mostRecentlySelectedEntityName: entityName })
 
-    @action.bound onSearchKeyDown(e: React.KeyboardEvent<HTMLElement>): void {
-        if (e.key === "Enter" && this.searchResults.length > 0) {
-            this.onSelect(this.searchResults[0].name)
-            this.searchInput = ""
-        }
+        this.clearSearchInput()
     }
 
     @action.bound onClear(): void {
-        this.selectionArray.clearSelection()
+        const { partitionedVisibleEntities: visibleEntities } = this
+        this.selectionArray.deselectEntities(visibleEntities.selected)
     }
```

Comment:
Okay this is a bit of a weird case, but:
I would suggest that "Clear" clears all entities (not just visible ones) if the search input is empty.
Otherwise, there are cases where one can never clear out all entities, because some entities are selected via the URL that are not available in the chart.
See http://staging-site-entity-selector-redesign/grapher/homicide-rates-across-western-europe?country=~England, for example.